d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ClusterFailoutBucketConfig.java

```diff
@@ -0,0 +1,131 @@
+  private final static String ONLINE_AT_PROPERTY = "onlineAt";
+
+  public ClusterFailoutBucketConfig(String fabricUrn,
+      String partition,
+      Integer bucketOfflineRange,
+      Date offlineAt,
+      Date onlineAt)
+  {
+    _fabricUrn = fabricUrn;
+    _partition = partition;
+    _bucketOfflineRange = bucketOfflineRange;
+    _offlineAt = offlineAt;
+    _onlineAt = onlineAt;
+  }
+
+  public static ClusterFailoutBucketConfig createFromMap(Map<String, Object> configMap) {
+    ClusterFailoutBucketConfig clusterFailoutBucketConfig = null;
+    try {
+      clusterFailoutBucketConfig = new ClusterFailoutBucketConfig((String) configMap.get(ClusterFailoutBucketConfig.FABRIC_URN_PROPERTY),
+          (String) configMap.get(ClusterFailoutBucketConfig.PARTITION_PROPERTY), (Integer) configMap.get(ClusterFailoutBucketConfig.BUCKET_OFFLINE_RANGE_PROPERTY),
+          (Date) configMap.get(ClusterFailoutBucketConfig.OFFLINE_AT_PROPERTY), (Date) configMap.get(ClusterFailoutBucketConfig.ONLINE_AT_PROPERTY));
+    }
+    catch(ClassCastException e) {
+      // Return value will be null if cast failed.
+    }
+    catch (NullPointerException e) {
+      // return value will be null if a key is missing.
+    }
+    return clusterFailoutBucketConfig;
+  }
+
+  public String getFabricUrn() {
+    return _fabricUrn;
+  }
+  public String getPartition() {
+    return _partition;
+  }
+  public Integer getBucketOfflineRange() {
+    return _bucketOfflineRange;
+  }
+  public Date getOfflineAt() {
+    return _offlineAt;
+  }
+  public Date getOnlineAt() {
+    return _onlineAt;
+  }
+
+
+  @Override
+  public String toString() {
+    return "ClusterFailoutBucketConfig [_fabricUrn=" + _fabricUrn
+        + ", partition=" + _partition
+        + ", bucketOfflineRange=" + _bucketOfflineRange
+        + ", offlineAt=" + _offlineAt
+        + ", onlineAt=" + _onlineAt
+        + "]";
+  }
+
+  @Override
+  public int hashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + _fabricUrn.hashCode();
+    result = prime * result + _partition.hashCode();
+    result = prime * result + _bucketOfflineRange.hashCode();
+    result = prime * result + _offlineAt.hashCode();
+    result = prime * result + _onlineAt.hashCode();
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj)
+  {
+    if (this == obj)
+      return true;
+    if (obj == null)
+      return false;
+    if (getClass() != obj.getClass())
+      return false;
+
+    ClusterFailoutBucketConfig other = (ClusterFailoutBucketConfig) obj;
+    if (!_fabricUrn.equals(other.getFabricUrn()))
+      return false;
+    if (!_partition.equals(other.getPartition()))
+      return false;
+    if (!_bucketOfflineRange.equals(other.getBucketOfflineRange()))
+      return false;
+    if (!_offlineAt.equals(other.getOfflineAt()))
+      return false;
+    if (!_onlineAt.equals(other.getOnlineAt()))
+      return false;
+    return true;
```

Comment:
nit: we can simplify to `return _fabricUrn.equals(other.getFabricUrn()) && _partition.equals(other.getPartition()) ...`